File: web/src/__tests__/dsl/directives/DirectiveFinder.test.ts
import { describe, it, expect } from 'vitest'
import { findDirectives, containsDirectives, directiveKey, startOffsetFromKey, hashDirective } from '../../../dsl/directives/DirectiveFinder'
import { executeDirective, executeAllDirectives } from '../../../dsl/directives/DirectiveExecutor'
import { DirectiveWarningType, directiveResultToValue, directiveResultToDisplayString, directiveResultWarning } from '../../../dsl/directives/DirectiveResult'
import { numberVal, stringVal } from '../../../dsl/runtime/DslValue'

describe('directiveKey', () => {
  it('creates key from lineId and offset', () => {
    expect(directiveKey('test-line', 5)).toBe('test-line:5')
    expect(directiveKey('test-line', 12)).toBe('test-line:12')
  })

  it('creates key with provided lineId', () => {
    expect(directiveKey('abc123', 5)).toBe('abc123:5')
    expect(directiveKey('note-xyz', 12)).toBe('note-xyz:12')
  })
})

describe('startOffsetFromKey', () => {
  it('extracts offset from noteId-based key', () => {
    expect(startOffsetFromKey('abc123:5')).toBe(5)
  })

  it('extracts offset from lineId-based key', () => {
    expect(startOffsetFromKey('test-line:12')).toBe(12)
  })

  it('returns undefined for invalid key', () => {
    expect(startOffsetFromKey('nocolon')).toBeUndefined()
  })

  it('returns undefined for non-numeric offset', () => {
    expect(startOffsetFromKey('abc:xyz')).toBeUndefined()
  })

  it('handles zero offset', () => {
    expect(startOffsetFromKey('test-line:0')).toBe(0)
    expect(startOffsetFromKey('note:0')).toBe(0)
  })
})

describe('findDirectives', () => {
  it('finds single directive', () => {
    const result = findDirectives('hello [world] there')
    expect(result).toHaveLength(1)
    expect(result[0]!.sourceText).toBe('[world]')
    expect(result[0]!.startOffset).toBe(6)
    expect(result[0]!.endOffset).toBe(13)
  })

  it('finds multiple directives', () => {
    const result = findDirectives('[foo] and [bar]')
    expect(result).toHaveLength(2)
    expect(result[0]!.sourceText).toBe('[foo]')
    expect(result[1]!.sourceText).toBe('[bar]')
  })

  it('returns empty for no directives', () => {
    const result = findDirectives('no directives here')
    expect(result).toHaveLength(0)
  })

  it('returns empty for empty content', () => {
    const result = findDirectives('')
    expect(result).toHaveLength(0)
  })

  it('handles nested brackets for bracket body syntax', () => {
    const result = findDirectives('[once[inner]]')
    expect(result).toHaveLength(1)
    expect(result[0]!.sourceText).toBe('[once[inner]]')
    expect(result[0]!.startOffset).toBe(0)
    expect(result[0]!.endOffset).toBe(13)
  })

  it('handles deeply nested brackets', () => {
    const result = findDirectives('[a[b[c]]]')
    expect(result).toHaveLength(1)
    expect(result[0]!.sourceText).toBe('[a[b[c]]]')
  })

  it('tracks correct offsets', () => {
    const result = findDirectives('abc [dir1] def [dir2]')
    expect(result).toHaveLength(2)
    expect(result[0]!.startOffset).toBe(4)
    expect(result[0]!.endOffset).toBe(10)
    expect(result[1]!.startOffset).toBe(15)
    expect(result[1]!.endOffset).toBe(21)
  })

  it('handles directive at start of line', () => {
    const result = findDirectives('[start] rest')
    expect(result).toHaveLength(1)
    expect(result[0]!.startOffset).toBe(0)
  })

  it('handles directive at end of line', () => {
    const result = findDirectives('rest [end]')
    expect(result).toHaveLength(1)
    expect(result[0]!.endOffset).toBe(10)
  })

  it('handles adjacent directives', () => {
    const result = findDirectives('[first][second]')
    expect(result).toHaveLength(2)
    expect(result[0]!.sourceText).toBe('[first]')
    expect(result[1]!.sourceText).toBe('[second]')
  })

  it('ignores unclosed bracket', () => {
    const result = findDirectives('[unclosed')
    expect(result).toHaveLength(0)
  })

  it('handles multiple directives on same line with text between', () => {
    const result = findDirectives('text [a] more [b] end')
    expect(result).toHaveLength(2)
    expect(result[0]!.sourceText).toBe('[a]')
    expect(result[0]!.startOffset).toBe(5)
    expect(result[1]!.sourceText).toBe('[b]')
    expect(result[1]!.startOffset).toBe(14)
  })

  it('handles directives spanning bracket patterns', () => {
    const result = findDirectives('[outer [inner] more]')
    expect(result).toHaveLength(1)
    expect(result[0]!.sourceText).toBe('[outer [inner] more]')
  })
})

describe('containsDirectives', () => {
  it('returns true when content has directives', () => {
    expect(containsDirectives('[hello]')).toBe(true)
    expect(containsDirectives('before [dir] after')).toBe(true)
  })

  it('returns false when content has no directives', () => {
    expect(containsDirectives('no brackets here')).toBe(false)
    expect(containsDirectives('')).toBe(false)
  })

  it('returns false for unclosed brackets', () => {
    expect(containsDirectives('[unclosed')).toBe(false)
  })
})

describe('hashDirective', () => {
  it('produces consistent hash for same input', async () => {
    const hash1 = await hashDirective('[test]')
    const hash2 = await hashDirective('[test]')
    expect(hash1).toBe(hash2)
  })

  it('produces different hash for different input', async () => {
    const hash1 = await hashDirective('[test1]')
    const hash2 = await hashDirective('[test2]')
    expect(hash1).not.toBe(hash2)
  })

  it('returns hex string', async () => {
    const hash = await hashDirective('[test]')
    expect(hash).toMatch(/^[0-9a-f]+$/)
    expect(hash.length).toBe(64) // SHA-256 = 32 bytes = 64 hex chars
  })
})

describe('executeDirective', () => {
  it('executes number directive successfully', () => {
    const result = executeDirective('[42]', [], null)
    expect(result.error).toBeNull()
    expect(result.result).not.toBeNull()
    const val = directiveResultToValue(result)
    expect(val).toEqual(numberVal(42))
  })

  it('executes string directive successfully', () => {
    const result = executeDirective('["hello"]', [], null)
    expect(result.error).toBeNull()
    const val = directiveResultToValue(result)
    expect(val).toEqual(stringVal('hello'))
  })

  it('returns error for invalid directive', () => {
    const result = executeDirective('[invalid@syntax]', [], null)
    expect(result.error).not.toBeNull()
    expect(result.error).toContain('error')
  })

  it('returns error for empty directive', () => {
    const result = executeDirective('[]', [], null)
    expect(result.error).not.toBeNull()
  })

  it('returns error for unclosed string', () => {
    const result = executeDirective('["unclosed]', [], null)
    expect(result.error).not.toBeNull()
  })
})

describe('executeAllDirectives', () => {
  it('returns results for all directives', () => {
    const { results } = executeAllDirectives('First [42] second ["hello"]', [], null)
    expect(results.size).toBe(2)
  })

  it('handles mixed success and error', () => {
    const { results } = executeAllDirectives('[42] and [invalid@]', [], null)
    expect(results.size).toBe(2)

    const values = [...results.values()]
    const success = values.find((r) => r.error === null)
    expect(success).toBeDefined()
    expect(directiveResultToValue(success!)).toEqual(numberVal(42))

    const error = values.find((r) => r.error !== null)
    expect(error).toBeDefined()
  })

  it('returns empty map for no directives', () => {
    const { results } = executeAllDirectives('No directives here', [], null)
    expect(results.size).toBe(0)
  })
})

describe('no-effect warnings', () => {
  it('implicit lambda at top level returns warning', () => {
    const result = executeDirective('[[i]]', [], null)
    expect(result.error).toBeNull()
    expect(result.warning).toBe(DirectiveWarningType.NO_EFFECT_LAMBDA)
  })

  it('implicit lambda with body at top level returns warning', () => {
    const result = executeDirective('[[i.path]]', [], null)
    expect(result.error).toBeNull()
    expect(result.warning).toBe(DirectiveWarningType.NO_EFFECT_LAMBDA)
  })

  it('pattern at top level returns warning', () => {
    const result = executeDirective('[pattern(digit*4)]', [], null)
    expect(result.error).toBeNull()
    expect(result.warning).toBe(DirectiveWarningType.NO_EFFECT_PATTERN)
  })

  it('warning display message is descriptive', () => {
    expect(DirectiveWarningType.NO_EFFECT_LAMBDA).toBe('Uncalled lambda has no effect')
    expect(DirectiveWarningType.NO_EFFECT_PATTERN).toBe('Unused pattern has no effect')
  })

  it('warning result toDisplayString shows warning message', () => {
    const result = directiveResultWarning(DirectiveWarningType.NO_EFFECT_LAMBDA)
    const display = directiveResultToDisplayString(result)
    expect(display).toContain('Warning')
    expect(display).toContain('Uncalled lambda')
  })
})

File: app/src/test/java/org/alkaline/taskbrain/dsl/directives/DirectiveFinderTest.kt
package org.alkaline.taskbrain.dsl.directives

import org.alkaline.taskbrain.dsl.runtime.values.NumberVal
import org.alkaline.taskbrain.dsl.runtime.values.StringVal
import org.junit.Assert.*
import org.junit.Test

class DirectiveFinderTest {

    private companion object {
        const val TEST_LINE_ID = "test-line"
    }

    // region Finding directives

    @Test
    fun `finds single directive in content`() {
        val content = "Some text [42] more text"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(1, directives.size)
        assertEquals("[42]", directives[0].sourceText)
        assertEquals(10, directives[0].startOffset)
        assertEquals(14, directives[0].endOffset)
    }

    @Test
    fun `finds multiple directives in content`() {
        val content = "[1] and [2] and [3]"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(3, directives.size)
        assertEquals("[1]", directives[0].sourceText)
        assertEquals("[2]", directives[1].sourceText)
        assertEquals("[3]", directives[2].sourceText)
    }

    @Test
    fun `finds string directive`() {
        val content = "Result: [\"hello\"]"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(1, directives.size)
        assertEquals("[\"hello\"]", directives[0].sourceText)
    }

    @Test
    fun `returns empty list for no directives`() {
        val content = "No directives here"
        val directives = DirectiveFinder.findDirectives(content)

        assertTrue(directives.isEmpty())
    }

    @Test
    fun `containsDirectives returns true when directives exist`() {
        assertTrue(DirectiveFinder.containsDirectives("Has [42] directive"))
    }

    @Test
    fun `containsDirectives returns false when no directives`() {
        assertFalse(DirectiveFinder.containsDirectives("No directives"))
    }

    @Test
    fun `finds directive at start of content`() {
        val content = "[42] at start"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(1, directives.size)
        assertEquals(0, directives[0].startOffset)
    }

    @Test
    fun `finds directive at end of content`() {
        val content = "at end [42]"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(1, directives.size)
        assertEquals(7, directives[0].startOffset)
        assertEquals(11, directives[0].endOffset)
    }

    @Test
    fun `handles empty content`() {
        val directives = DirectiveFinder.findDirectives("")
        assertTrue(directives.isEmpty())
    }

    // endregion

    // region Nested brackets (Milestone 8 - lambdas)

    @Test
    fun `finds directive with nested brackets`() {
        val content = "Test [[i]] here"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(1, directives.size)
        assertEquals("[[i]]", directives[0].sourceText)
        assertEquals(5, directives[0].startOffset)
        assertEquals(10, directives[0].endOffset)
    }

    @Test
    fun `finds directive with deeply nested brackets`() {
        val content = "[[matches(i.path, pattern(digit*4))]]"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(1, directives.size)
        assertEquals(content, directives[0].sourceText)
    }

    @Test
    fun `finds multiple directives with nested brackets`() {
        val content = "[[i.path]] and [[i.name]]"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(2, directives.size)
        assertEquals("[[i.path]]", directives[0].sourceText)
        assertEquals("[[i.name]]", directives[1].sourceText)
    }

    @Test
    fun `finds mixed nested and simple directives`() {
        val content = "[42] then [[i]] then [\"hello\"]"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(3, directives.size)
        assertEquals("[42]", directives[0].sourceText)
        assertEquals("[[i]]", directives[1].sourceText)
        assertEquals("[\"hello\"]", directives[2].sourceText)
    }

    @Test
    fun `handles unmatched opening bracket`() {
        val content = "Test [unclosed"
        val directives = DirectiveFinder.findDirectives(content)

        assertTrue(directives.isEmpty())
    }

    @Test
    fun `handles unmatched nested bracket`() {
        val content = "Test [[unclosed]"
        val directives = DirectiveFinder.findDirectives(content)

        // The outer bracket is unmatched, so no directive found
        assertTrue(directives.isEmpty())
    }

    // endregion

    // region Executing directives

    @Test
    fun `executes number directive successfully`() {
        val execResult = DirectiveFinder.executeDirective("[42]")

        assertNull(execResult.result.error)
        assertNotNull(execResult.result.result)
        assertEquals(42.0, (execResult.result.toValue() as NumberVal).value, 0.0)
    }

    @Test
    fun `executes string directive successfully`() {
        val execResult = DirectiveFinder.executeDirective("[\"hello\"]")

        assertNull(execResult.result.error)
        assertEquals("hello", (execResult.result.toValue() as StringVal).value)
    }

    @Test
    fun `returns error for invalid directive`() {
        val execResult = DirectiveFinder.executeDirective("[invalid@syntax]")

        assertNotNull(execResult.result.error)
        assertTrue(execResult.result.error!!.contains("error"))
    }

    @Test
    fun `returns error for empty directive`() {
        val execResult = DirectiveFinder.executeDirective("[]")

        assertNotNull(execResult.result.error)
    }

    @Test
    fun `returns error for unclosed string`() {
        val execResult = DirectiveFinder.executeDirective("[\"unclosed]")

        assertNotNull(execResult.result.error)
        assertTrue(execResult.result.error!!.contains("Lexer error"))
    }

    // endregion

    // region Execute all directives

    @Test
    fun `executeAllDirectives returns results for all directives`() {
        val content = "First [42] second [\"hello\"]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)

        assertEquals(2, results.size)
    }

    @Test
    fun `executeAllDirectives handles mixed success and error`() {
        val content = "[42] and [invalid@]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)

        assertEquals(2, results.size)

        // Find the successful one
        val successResult = results.values.find { it.error == null }
        assertNotNull(successResult)
        assertEquals(42.0, (successResult!!.toValue() as NumberVal).value, 0.0)

        // Find the failed one
        val errorResult = results.values.find { it.error != null }
        assertNotNull(errorResult)
    }

    @Test
    fun `executeAllDirectives returns empty map for no directives`() {
        val results = DirectiveFinder.executeAllDirectives("No directives here", TEST_LINE_ID)
        assertTrue(results.isEmpty())
    }

    // endregion

    // region No-effect warnings (Milestone 8)

    @Test
    fun `lambda at top level returns warning`() {
        val execResult = DirectiveFinder.executeDirective("[[i]]")

        assertNull(execResult.result.error)
        assertNotNull(execResult.result.warning)
        assertEquals(DirectiveWarningType.NO_EFFECT_LAMBDA, execResult.result.warning)
        assertTrue(execResult.result.hasWarning)
    }

    @Test
    fun `lambda with body at top level returns warning`() {
        val execResult = DirectiveFinder.executeDirective("[[i.path]]")

        assertNull(execResult.result.error)
        assertEquals(DirectiveWarningType.NO_EFFECT_LAMBDA, execResult.result.warning)
    }

    @Test
    fun `pattern at top level returns warning`() {
        val execResult = DirectiveFinder.executeDirective("[pattern(digit*4)]")

        assertNull(execResult.result.error)
        assertNotNull(execResult.result.warning)
        assertEquals(DirectiveWarningType.NO_EFFECT_PATTERN, execResult.result.warning)
    }

    @Test
    fun `warning display message is descriptive`() {
        assertEquals("Uncalled lambda has no effect", DirectiveWarningType.NO_EFFECT_LAMBDA.displayMessage)
        assertEquals("Unused pattern has no effect", DirectiveWarningType.NO_EFFECT_PATTERN.displayMessage)
    }

    @Test
    fun `warning result toDisplayString shows warning message`() {
        val result = DirectiveResult.warning(DirectiveWarningType.NO_EFFECT_LAMBDA)

        assertTrue(result.toDisplayString().contains("Warning"))
        assertTrue(result.toDisplayString().contains("Uncalled lambda"))
    }

    // endregion

    // region Directive hashing

    @Test
    fun `same directive text produces same hash`() {
        val content = "[42] and [42]"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(2, directives.size)
        assertEquals(directives[0].hash(), directives[1].hash())
    }

    @Test
    fun `different directive text produces different hash`() {
        val content = "[42] and [43]"
        val directives = DirectiveFinder.findDirectives(content)

        assertEquals(2, directives.size)
        assertNotEquals(directives[0].hash(), directives[1].hash())
    }

    @Test
    fun `hash returns hex string`() {
        val directive = DirectiveFinder.findDirectives("[test]")[0]
        val hash = directive.hash()
        assertTrue(hash.matches(Regex("^[0-9a-f]+$")))
    }

    // endregion

    // region directiveKey

    @Test
    fun `directiveKey produces lineId-based key`() {
        assertEquals("abc123:15", DirectiveFinder.directiveKey("abc123", 15))
    }

    @Test
    fun `directiveKey with test lineId produces expected key`() {
        assertEquals("test-line:15", DirectiveFinder.directiveKey(TEST_LINE_ID, 15))
    }

    @Test
    fun `directiveKey formats correctly`() {
        assertEquals("note1:5", DirectiveFinder.directiveKey("note1", 5))
    }

    // endregion

    // region startOffsetFromKey

    @Test
    fun `startOffsetFromKey extracts offset from lineId key`() {
        assertEquals(15, DirectiveFinder.startOffsetFromKey("abc123:15"))
    }

    @Test
    fun `startOffsetFromKey extracts offset from test lineId key`() {
        assertEquals(20, DirectiveFinder.startOffsetFromKey("test-line:20"))
    }

    @Test
    fun `startOffsetFromKey returns null for non-numeric`() {
        assertNull(DirectiveFinder.startOffsetFromKey("bad:key:abc"))
    }

    @Test
    fun `startOffsetFromKey handles zero offset`() {
        assertEquals(0, DirectiveFinder.startOffsetFromKey("note1:0"))
        assertEquals(0, DirectiveFinder.startOffsetFromKey("test-line:0"))
    }

    @Test
    fun `startOffsetFromKey returns null for key with no colon`() {
        assertNull(DirectiveFinder.startOffsetFromKey("nocolon"))
    }

    // endregion
}
